Virtuozzo kernel compat: handle kernel DIRECT_IO via ITERs while VFS_RW_ITERATE are unsupported

### System information
<!--  add version after "|" character -->
Type | Version/Name
 --- | ---
Distribution Name	| Virtuozzo
Distribution Version	| 7.5 (Update 15)
Linux Kernel	| 3.10.0-1127.18.2.vz7.163.46
Architecture	| x86_64
ZFS Version	| zfs-0.8.5 and zfs-2.0.0
SPL Version	| spl-0.7.13
<!--
Commands to find ZFS/SPL versions:
modinfo zfs | grep -iw version
modinfo spl | grep -iw version
-->

### Describe the problem you're observing
We, Virtuozzo, got a report that zfs module fails to compile for Virtuozzo kernel 3.10.0-1127.18.2.vz7.163.46
http://repo.virtuozzo.com/vz/releases/7.5.0-586/x86_64/os/Packages/v/vzkernel-3.10.0-1127.18.2.vz7.163.46.x86_64.rpm

while it compiled fine for previous Virtuozzo kernel 3.10.0-1127.8.2.vz7.151.14
http://repo.virtuozzo.com/vz/releases/7.5.0-586/x86_64/os/Packages/v/vzkernel-3.10.0-1127.8.2.vz7.151.14.x86_64.rpm

https://bugs.openvz.org/browse/OVZ-7243

### Describe how to reproduce the problem
Get an installation of Virtuozzo 7.5 (optionally just install vzkernel and vzkernel-devel packages for version 3.10.0-1127.18.2.vz7.163.46 from http://repo.virtuozzo.com/vz/releases/7.5.0-586/x86_64/os/Packages/v/ in a, say, CentOS 7 environment) and try to compile zfs module.
You will get the following compilation error:

```
zfs-kmod-2.0.0/_kmod_build_3.10.0-1127.18.2.vz7.163.46/../zfs-2.0.0/module/zfs/../os/linux/zfs/zpl_file.c:502:2:
error: #error "Unknown direct IO interface"
zfs-kmod-2.0.0/_kmod_build_3.10.0-1127.18.2.vz7.163.46/../zfs-2.0.0/module/zfs/../os/linux/zfs/zpl_file.c:1026:15:
error: 'zpl_direct_IO' undeclared here (not in a function)
  .direct_IO = zpl_direct_IO,
```

### Whats is the reason?
Virtuozzo kernel does not have vfs rw iterators (**struct file_operations :: .read_iter()/.write_iter()**) compatible with mainstream version, so zfs correctly does not detect HAVE_VFS_RW_ITERATE define.

Previously (for example in vzkernel vz7.151.14) **struct address_space_operations :: .direct_IO()** worked via IOVECs and zfs correctly detected HAVE_VFS_DIRECT_IO_IOVEC and compilation worked fine, **.direct_IO()** definition via iovec was used.

zpl_file.c:
```
#if defined(HAVE_VFS_RW_ITERATE)

#if defined(HAVE_VFS_DIRECT_IO_ITER)
lalala
#elif defined(HAVE_VFS_DIRECT_IO_ITER_OFFSET)
lalala
#elif defined(HAVE_VFS_DIRECT_IO_ITER_RW_OFFSET)
lalala            // we should be here in vz7.163.x
#else
#error "Unknown direct IO interface"
#endif

#else

#if defined(HAVE_VFS_DIRECT_IO_IOVEC)
lalala            // we are here in vz7.151.14
#else
#error "Unknown direct IO interface"
#endif
#endif /* HAVE_VFS_RW_ITERATE */
```

But recently we have changed **struct address_space_operations :: .direct_IO()** to use iterators and now we came to a situation that zfs configurator
* does not detect HAVE_VFS_RW_ITERATE (correctly)
* does detect HAVE_VFS_DIRECT_IO_ITER_RW_OFFSET (correctly)
* now .direct_IO() definition is found.

i suggest to add a .direct_IO() definition for the case HAVE_VFS_DIRECT_IO_ITER_RW_OFFSET + no HAVE_VFS_RW_ITERATE.

Will send the pull request soon.

**Note:** even with this patch in zfs, it won't compile for kernel 3.10.0-1127.18.2.vz7.163.46, another kernel patch is to be applied first (we will apply it soon): https://lists.openvz.org/pipermail/devel/2020-December/076222.html
If you want to test the compilation right now, until we make an official kernel with this patch, you can take a test kernel at
http://fe.virtuozzo.com/ceacebe1809ed2631696da3ea06c3853/

> error: commit subject over 72 characters
> make: *** [commitcheck] Error 1

Should i fix it? i thought the rule about string length does not apply to the subject line.

--- a/module/os/linux/zfs/zpl_file.c
+++ b/module/os/linux/zfs/zpl_file.c
@@ -463,7 +463,7 @@ zpl_direct_IO(int rw, struct kiocb *kiocb, struct iov_iter *iter, loff_t pos)
 #error "Unknown direct IO interface"
 #endif
 
-#else
+#else /* HAVE_VFS_RW_ITERATE */
 
 #if defined(HAVE_VFS_DIRECT_IO_IOVEC)
 static ssize_t
@@ -474,6 +474,19 @@ zpl_direct_IO(int rw, struct kiocb *kiocb, const struct iovec *iov,
 		return (zpl_aio_write(kiocb, iov, nr_segs, pos));
 	else
 		return (zpl_aio_read(kiocb, iov, nr_segs, pos));
+}
+#elif defined(HAVE_VFS_DIRECT_IO_ITER_RW_OFFSET)
+static ssize_t
+zpl_direct_IO(int rw, struct kiocb *kiocb, struct iov_iter *iter, loff_t pos)
+{
+	const struct iovec *iovp = iov_iter_iovec(iter);
+	unsigned long nr_segs = iter->nr_segs;
+
+	ASSERT3S(pos, ==, kiocb->ki_pos);
+	if (rw == WRITE)
+		return (zpl_aio_write(kiocb, iovp, nr_segs, pos));
+	else
+		return (zpl_aio_read(kiocb, iovp, nr_segs, pos));
 }
 #else
 #error "Unknown direct IO interface"

